plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Currency.hs

```diff
@@ -0,0 +1,171 @@
+
+-- | The 'Value' forged by the 'curValidator' contract
+forgedValue :: Currency -> Value
+forgedValue cur = 
+    let a = plcCurrencySymbol (curAddress cur)
+        t = curRefTokenName cur
+        i = curRefAmount cur
+    in
+        $$(Value.singleton) a t i
+
+-- | @forge c n@ forges @n@ units of a currency called @c@ and
+--   pays them to a public key address owned by the wallet.
+forge :: (WalletAPI m, WalletDiagnostics m) => String -> Int -> m Currency
+forge nm amount = do
+    pk <- WAPI.ownPubKey
+
+    -- 1. We need to create the reference transaction output using the 
+    --    'PublicKey' contract. That way we get an output that behaves
+    --    like a normal public key output, but is not selected by the
+    --    wallet during coin selection. This ensures that the output still 
+    --    exists when we spend it in our forging transaction.
+    (refAddr, refTxIn) <- PK.lock pk (Ada.adaValueOf 1)
+
+    let
+         -- With that we can define the currency
+        theCurrency = mkCurrency (txInRef refTxIn) amount nm
+        curAddr     = curAddress theCurrency
+        forgedVal   = forgedValue theCurrency
+        oneOrMore   = WAPI.intervalFrom $ Ada.adaValueOf 1
+
+        -- trg1 fires when 'refTxIn' can be spent by our forging transaction
+        trg1 = fundsAtAddressT refAddr oneOrMore
+        
+        -- trg2 fires when the pay-to-script output locked by 'curValidator' 
+        -- is ready to be spent.
+        trg2 = fundsAtAddressT curAddr oneOrMore
+
+        -- The 'forge_' action creates a transaction that spends the contract
+        -- output, forging the currency in the process.
+        forge_ :: (WalletAPI m, WalletDiagnostics m) => m ()
+        forge_ = do
+            ownOutput <- WAPI.ownPubKeyTxOut (forgedVal <> Ada.adaValueOf 2)
+            am <- WAPI.watchedAddresses
+
+            let inputs' = am ^. at curAddr . to (Map.toList . fromMaybe Map.empty)
+                con (r, _) = scriptTxIn r (curValidator theCurrency) curRedeemer
+                ins        = con <$> inputs'
+
+            let tx = Ledger.Tx
+                        { txInputs = Set.fromList (refTxIn:ins)
+                        , txOutputs = [ownOutput]
+                        , txForge = forgedVal
+                        , txFee   = Ada.zero
+                        , txValidRange = defaultSlotRange
+                        , txSignatures = Map.empty
+                        }
+
+            WAPI.logMsg $ Text.pack $ "Forging transaction: " <> show (Ledger.hashTx tx)
+            WAPI.signTxAndSubmit_  tx
+
+    -- 2. We start watching the contract address, ready to forge
+    --    our currency once the monetary policy script has been
+    --    placed on the chain.
+    registerOnce trg2 (EventHandler $ const forge_)
+
+    -- 3. When trg1 fires we submit a transaction that creates a 
+    --    pay-to-script output locked by the monetary policy
+    registerOnce trg1 (EventHandler $ const $ do

```

Comment:
It's annoying that we have to do this, but if we submit the transactions together there's no guarantee they will happen in the right order. It would be nice to be able to submit a transaction bundle that would either be validated all together or not at all...

Reply:
like a transaction-transaction?

Reply:
Maybe. I don't know what I want yet. In terms of what I was talking about yesterday, we're doing two adjacent chunks of logic "within" the same ownership, so it seems like you should be able to glue them together in the same way that we do for normal transactions. But possibly I'm just getting overly attached to the "contiguous block of logic with no ownership changes = transaction" idea.